api/src/main/java/io/minio/MinioClient.java

```diff
@@ -2294,6 +2297,182 @@ public void copyObject(String bucketName, String objectName, String destBucketNa
+          InvalidKeyException, NoResponseException, XmlPullParserException, ErrorResponseException,
+          InternalException, InvalidArgumentException {
+
+    if (listSourceObjects.size() < 1 || listSourceObjects.size() > MAX_MULTIPART_COUNT ) {
+      throw new InvalidArgumentException("There must be as least one and up to 10000 source objects.");
+    }
+    int i = 0;
+    long totalSize = 0;
+    for (CopySource sourceObject : listSourceObjects) {
+      i++;
+      ObjectStat os = statObject(sourceObject.bucketName(), sourceObject.objectName());
+      // Error out if client side encryption is used in this source object when
+      // more than one source objects are given.
+      if ((listSourceObjects.size() > 1 )  && sourceObject.headers().get("x-amz-meta-x-amz-key") !=  null) {
+        throw new InvalidArgumentException("Client side encryption is used in source object"
+                + sourceObject.bucketName() + sourceObject.objectName());
+      }
+
+      // Check if a segment is specified is within object bounds
+      if ( sourceObject.length() > os.length()) {
+        throw new InvalidArgumentException("Source specified length " + sourceObject.length() + " is more than  "
+                +  "size of Object " + os.length());
+      }
+
+      if ( os.length() < MIN_MULTIPART_SIZE && i < listSourceObjects.size() - 1 ) {
+        throw new InvalidArgumentException(" Source Object size " + os.length() + " is too small "
+         +  " and " + i + " it is not the last part");
+      }
+
+      totalSize = totalSize + os.length();
+      if ( totalSize > MAX_OBJECT_SIZE ) {
+        throw new InvalidArgumentException(" Cannot compose an object of size more than " + MAX_OBJECT_SIZE );
+      }
+    }
+
+    // Single source object case (i.e. when only one source is
+    // involved, it is being copied wholly and at most 5GiB in
+    // size, emptyfiles are also supported).
+    if ( listSourceObjects.size() == 1 ) {
+      CopySource  sourceObject = listSourceObjects.get(0);
+      copyObject(sourceObject.bucketName(), sourceObject.objectName(),
+              destinationBucketName, destintionObjectName);
+    }
+
+    if ( userMetaData != null  && listSourceObjects.size() == 1  ) {
+      CopySource  source = listSourceObjects.get(0);
+      userMetaData.putAll(source.headers());
+    }
+
+    // Now, handle multipart-copy cases.
+    Map<String, String> headerMap = new HashMap<>();
+    if ( userMetaData != null) {
+      headerMap.putAll(userMetaData);
+    }
+    if (sse != null) {
+      sse.marshal(headerMap);
+    }
+    String uploadId = initMultipartUpload(destinationBucketName,destintionObjectName , headerMap);
+    int partNumber = 1;
+    Part[] totalPart = new Part[listSourceObjects.size()];
+
+    for (CopySource source : listSourceObjects) {
+      Map<String, String> sourceMap  = source.headers();
+      ObjectStat os = statObject(source.bucketName(), source.objectName());
```

Comment:
Why is it done once again?  You could merge or do checks/constructions in only one loop.

Reply:
All the validations are done in the first for loop. Lets say we have to compose object from 800 objects. the validations fails for 799th object so all the parts copied are useless. So we first validate and then compose object. The same in done in `minio-go` sdk

Reply:
Yeah, unfortunately in the compose API, even if the request appears to be valid, the operation may not be feasible - some source objects may be larger than 5GB and may need to be broken up into multiple part uploads - this may in turn require more than 10K parts. For this reason a separate loop performs the validation and a second loop performs the upload. To ensure that the source objects do not change between the first and second loop, the second loop uses the Src-If-Match condition in the go-implementation, this should be done here too. This feature and all these implementation details were discussed with @abperiasamy when it was first implemented in Go (some customer had requested this feature at that time).

Reply:
@balamurugana  PTAL